### PR TITLE
feat: logging using tracing crate 

### PR DESCRIPTION
**Pull request by plaxi0s**
WIP #30 
Currently logs to stdout only. Will add a flag for logfile after this. 
Sample logging 

image
What do you think @tkellogg ?

### CODE REVIEW

**Comment by tkellogg** [comment]
wow! this is great. I haven't gotten a chance to comb through the code, so my only feedback is that I would rather the output be parseable JSON. I have dreams of using dura to do analytics on myself (not sure what, maybe like percent of time coding or something, first step is to capture parseable data)

**Comment by plaxi0s** [comment]
I feel what you are saying. My only concern is about logs related to application logic/errors. Do you want to club them with logs used for analytics?

**Comment by tkellogg** [comment]
aren't all logs for analytics? at my last several jobs we've turned to JSON logs for everything, because of how much automation you can add. Not everyone is a whiz at awk & grep

**Comment by plaxi0s** [comment]
yeah, you are right! I misunderstood what you were saying. We should go for it.

File: src/log.rs [hunk 1]
```diff
@@ -1,94 +1,46 @@
...
-        if !log.op.should_log() {
-            return
-        }
-
-        let _ = self.ensure_open();
-        if let Some(ref file) = self.file {
-            let writer = io::LineWriter::new(file);
-            if let Err(error) = serde_json::to_writer(writer, &log) {
-                eprintln!("Unable to write log: {:?}", error);
+impl Display for Operation {
```
Comment (tkellogg) [inline src/log.rs:29]: Oh goodness, you went through all this work to display it, and I'm telling you to do it in JSON, sorry 😊

File: src/poller.rs [hunk 2]
```diff
@@ -3,39 +3,54 @@ use std::process;
...
+use tracing::{error, info};
 
-use crate::snapshots;
 use crate::config::Config;
-use crate::log::{Log, Logger, Operation};
+use crate::log::Operation;
+use crate::snapshots;
 
-fn process_directory(path: &Path) -> Log {
+#[tracing::instrument]
```
Comment (tkellogg) [inline src/poller.rs:12]: What does this do? Looks neat

**Review by tkellogg — COMMENTED** [review]
This is a great change, thank you! But I would prefer it printed in JSON

File: src/poller.rs (hunk 2 above)
Comment (plaxi0s) [reply src/poller.rs:12]: This macro basically creates a new span each time a function is called.

File: src/log.rs (hunk 1 above)
Comment (plaxi0s) [reply src/log.rs:29]: No problem. Should have gone for JSON from the beginning.

**Comment by plaxi0s** [comment]
The logger now prints JSON to stdout. 
Sample  
image

**Comment by tkellogg** [comment]
looking at the image, i'm not sure that will parse. can you run `jq . the-log.json`? (might have to install jq, but it's super useful so you should do it anyway). i expected each output to be in its own line, but it doesn't have to be that way

**Comment by tkellogg** [comment]
I spun it up to look. A couple more issues:

1. It didn't `flush` the full field, so I could be waiting a long time until it parses correctly 
image
2. I don't see a timestamp
3. It has non-json in `stdout`, e.g. `pid: 38536`, so that would cause JSON parsers to choke.

**Comment by plaxi0s** [comment]
The logs are getting parsed correctly using `jq . log.json`. Added a timestamp to the log messages. Also each log message has its own line.

**Review by tkellogg — APPROVED** [review]

**Pull request merged**